pkg/test/framework/components/echo/echotest/run.go

```diff
@@ -15,31 +15,81 @@
 package echotest
 
 import (
+	"strings"
+
+	"istio.io/istio/pilot/pkg/networking/util"
 	"istio.io/istio/pkg/test/framework"
 	"istio.io/istio/pkg/test/framework/components/echo"
 )
 
 type (
-	perDeploymentTest func(t framework.TestContext, instances echo.Instances)
-	perInstanceTest   func(t framework.TestContext, src echo.Instance, dst echo.Instances)
+	perDeploymentTest  func(t framework.TestContext, instances echo.Instances)
+	perNDeploymentTest func(t framework.TestContext, deployments echo.Services)
+	perInstanceTest    func(t framework.TestContext, inst echo.Instance)
+
+	oneToOneTest func(t framework.TestContext, src echo.Instance, dst echo.Instances)
+	oneToNTest   func(t framework.TestContext, src echo.Instance, dsts echo.Services)
 )
 
-// Run will generate nested subtests for every instance in every deployment. The subtests will be nested including
-// the source service, source cluster and destination deployment. Example: `a/to_b/from_cluster-0`
-func (t *T) Run(testFn perInstanceTest) {
+// Run will generate and run one subtest to send traffic between each combination
+// of source instance to target deployment.
```

Comment:
Is it accurate to just say `source and target echo instance`? An echo instance is equivalent to a deployment.

Reply:
The correct description would be `source deployment -> target service`.

Reply:
Ah I see, `oneToOneTest` has `src echo.Instance, dst echo.Instances`. So `dst` in this case is assumed to have multiple `Instance`s for the same service (fqdn)? If so, perhaps it should be of type `Service` which is the same as `Instances` but has the requirement that each `Instance` is for the same fqdn.